### PR TITLE
consistent CI

### LINKED ISSUE #9766: consistent automated style, build, and test workflows
## workflow(s) triggered by pull request
- [x] `pre-commit`
  - [x] required to merge
  - [x] as a workflow
- [x] readthedocs build
  - [x] required to merge
- [x] check changelog
  - [ ] required to merge
  - [x] ensure news fragment exists for the triggering pull request (with `no-changelog-entry-needed` label to bypass)
  - [x] prevent manual changelog edits (with `allow-manual-changelog-edit` label to bypass)
- [x] build check
  - [ ] required to merge
- [x] unit tests on Linux Python 3.11
  - [x] required to merge
- [x] unit tests on Linux Python 3.12
  - [ ] required to merge
- [x] unit tests on Linux Python 3
  - [x] required to merge
- [x] unit tests on macOS Python 3
  - [ ] required to merge

## workflow triggered on schedule and / or label on pull request
- [x] scheduled weekly
- [x] triggerable on a pull request with the `run scheduled tests` label
- [ ] unit tests on macOS Python 3.11
- [x] unit tests on macOS Python 3.12
- [x] unit tests with development versions of dependencies (`devdeps`)
  - [x] Linux Python 3 `devdeps`
  - [x] macOS Python 3 `devdeps`

### PR DESCRIPTION
**Pull request by zacharyburnett**
closes #9766 

## Tasks
- [x] If you have a specific reviewer in mind, tag them.
- [ ] add a build milestone, i.e. `Build 12.0` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [ ] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [ ] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see [changelog readme](https://github.com/spacetelescope/jwst/blob/main/changes/README.rst) for instructions) 
  - [ ] update or add relevant tests
  - [ ] update relevant docstrings and / or `docs/` page
  - [ ] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [ ] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -1,9 +1,8 @@
 [tox]
 envlist =
-    check-dependencies
+    check-{dependencies,docs-links}
```
Comment (pllim) [inline tox.ini:3]: That is a lot of diff to shorten this file by one line... 😆

File: .github/workflows/ci_cron.yml [hunk 2]
```diff
@@ -32,10 +32,10 @@ concurrency:
 
 jobs:
   latest_crds_contexts:
-    if: (github.repository == 'spacetelescope/jwst' && (github.event_name == 'schedule' || github.event_name == 'workflow_dispatch' || contains(github.event.pull_request.labels.*.name, 'run scheduled tests')))
+    if: (github.repository == 'spacetelescope/jwst' && (github.event_name == 'schedule' || github.event_name == 'workflow_dispatch' || contains(github.event.pull_request.labels.*.name, 'run extra tests')))
```
Comment (pllim) [inline .github/workflows/ci_cron.yml:35]: Not sure if I like this as this involves messing with existing labels for no gain. What is your motivation here?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -32,14 +32,6 @@ concurrency:
   cancel-in-progress: true
 
 jobs:
-  pre-commit:
```
Comment (pllim) [inline .github/workflows/ci.yml:35]: No need for pre-commit anymore? Why? What is replacing it?

File: .github/workflows/changelog.yml [hunk 4]
```diff
@@ -14,7 +14,7 @@ concurrency:
   cancel-in-progress: true
 
 jobs:
-  check:
+  changelog-check:
```
Comment (pllim) [inline .github/workflows/changelog.yml:17]: This renaming requires a repo admin to change branch protection rule for little gain. Is this really necessary?

File: .github/workflows/ci_cron.yml [hunk 5]
```diff
@@ -59,10 +59,10 @@ jobs:
...
         results.xml
       envs: |
-        - macos: py312-xdist
+        - linux: check-docs-links
+        - macos: py311-xdist
           pytest-results-summary: true
-        - linux: py311-xdist
+        - macos: py312-xdist
           pytest-results-summary: true
         - linux: py3-xdist
```
Comment (braingram) [inline .github/workflows/ci_cron.yml]: Would you remove this? It's already run (on a schedule) in `ci.yml`:
https://github.com/spacetelescope/jwst/blob/fa200e7a3bf0161176ddddc14cd70525ff4d1210/.github/workflows/ci.yml#L75

**Comment by braingram** [comment]
I don't see any changes to `test_devdeps.yml`. The issue mentions devdeps as part of the "extra" tests. Does it make sense to combine that workflow tith `ci_cron.yml`?

File: .github/workflows/changelog.yml (hunk 4 above)
Comment (braingram) [reply .github/workflows/changelog.yml:17]: I'm in favor of the new name. `check` is pretty ambiguous.

File: .github/workflows/ci_cron.yml (hunk 2 above)
Comment (braingram) [reply .github/workflows/ci_cron.yml:35]: I'm guessing the idea here is that "scheduled" isn't very descriptive (and since the `ci.yml` jobs also run on a schedule "scheduled" is imprecise). I would like to propose something more descriptive then "extra" but I can't think of something that describes the "not normally run as part of the CI but run on a weekly basis jobs":
- mac tests for non-py3 version
- devdeps
- check-docs-links

any ideas?

Comment (zacharyburnett) [reply .github/workflows/ci_cron.yml:35]: `run additional tests`, `run unusual tests`, or maybe `run expensive tests`?

**Comment by zacharyburnett** [comment]
> I don't see any changes to `test_devdeps.yml`. The issue mentions devdeps as part of the "extra" tests. Does it make sense to combine that workflow tith `ci_cron.yml`?

good idea!

File: .github/workflows/ci.yml (hunk 3 above)
Comment (zacharyburnett) [reply .github/workflows/ci.yml:35]: I set up the `pre-commit.ci` bot, which runs on PRs automatically, so this should be offloaded there

File: .github/workflows/changelog.yml (hunk 4 above)
Comment (zacharyburnett) [reply .github/workflows/changelog.yml:17]: yeah when I originally made this workflow I assumed it would name itself `changelog / check`, but instead it's just named `check`

File: .github/workflows/ci_cron.yml (hunk 2 above)
Comment (braingram) [reply .github/workflows/ci_cron.yml:35]: Thanks for the suggestions. I'm pretty sure some of the tests I added to jwst would fall under the "unusual" umbrella :)

I'm warming up to "extra". Perhaps also rename `ci_cron.yml` to `ci_extra.yml` to help distinguish those tests.

Comment (zacharyburnett) [reply .github/workflows/ci_cron.yml:35]: I would prefer to have `ci_extra.yml` (or `tests_extra.yml`), but unfortunately GitHub tracks the history of the workflow by the filename, so if we rename it then it'll start a new sub-tab under the Actions tab with the same name, and that'll probably be really confusing for people

Comment (braingram) [reply .github/workflows/ci_cron.yml:35]: Screenshot 2025-08-25 at 4 11 40 PM

Are these the sub-tabs you're referring to? Since "extra" now had devdeps and what was previously "scheduled" would giving it a new name (something other than "test on schedule") help avoid confusion?

Also, does anyone watch/check these? If so perhaps they can weigh in on what to call this and if a renaming would be overly disruptive.

File: .github/workflows/changelog.yml (hunk 4 above)
Comment (pllim) [reply .github/workflows/changelog.yml:17]: OK then

File: .github/workflows/ci.yml (hunk 3 above)
Comment (pllim) [reply .github/workflows/ci.yml:35]: OK I updated OP to also close #9645

File: .github/workflows/ci_cron.yml (hunk 2 above)
Comment (pllim) [reply .github/workflows/ci_cron.yml:35]: This discussion is moot. Looks like you already deleted the old labels and added this new label.

Now the `main` workflows are broken unless we merge this PR. 🤷‍♀️

File: .github/workflows/ci_cron.yml [hunk 6]
```diff
@@ -59,10 +59,16 @@ jobs:
...
       envs: |
+        - linux: check-docs-links
+        - macos: py311-xdist
+          pytest-results-summary: true
         - macos: py312-xdist
           pytest-results-summary: true
-        - linux: py311-xdist
+        - linux: py3-stdevdeps-xdist
+          pytest-results-summary: true
+        - macos: py3-stdevdeps-xdist
```
Comment (pllim) [inline .github/workflows/ci_cron.yml:70]: Didn't @braingram say we should just merge stdevdeps into devdeps?

**Comment by pllim** [comment]
This PR now blocks all other open PRs because the branch protection rules were changed pre-maturely before this PR is merged.

Screenshot 2025-08-25 174353

**Comment by melanieclarke** [comment]
> This PR now blocks all other open PRs because the branch protection rules were changed pre-maturely before this PR is merged.

@zacharyburnett - can we revert this change for now, until this PR is merged?

File: .github/workflows/ci_cron.yml (hunk 2 above)
Comment (zacharyburnett) [reply .github/workflows/ci_cron.yml:35]: I apologize, I just renamed the label (it can be renamed at will until we settle on a name and all PRs that use it will be updated). If it makes more sense, I can rename the label back to the old name (`run scheduled tests`) and make a separate label with the new name (`run extra tests`) that we can play around with.

Labels also have an `id` attribute that doesn't change with the name, which would be better to use if we wanted to use a different name in the future, but I'm not sure exactly how to determine the ID of an arbitrary label from the web UI

**Comment by zacharyburnett** [comment]
> This PR now blocks all other open PRs because the branch protection rules were changed pre-maturely before this PR is merged.
> 
> Screenshot 2025-08-25 174353

> > This PR now blocks all other open PRs because the branch protection rules were changed pre-maturely before this PR is merged.
> 
> @zacharyburnett - can we revert this change for now, until this PR is merged?

sure thing, done

**Comment by melanieclarke** [comment]
I'm also seeing "Cannot update this protected ref" on this and other PRs. Might that also be branch protection related?

File: .github/workflows/ci_cron.yml (hunk 2 above)
Comment (zacharyburnett) [reply .github/workflows/ci_cron.yml:35]: > Screenshot 2025-08-25 at 4 11 40 PM
> Are these the sub-tabs you're referring to? Since "extra" now had devdeps and what was previously "scheduled" would giving it a new name (something other than "test on schedule") help avoid confusion?
> 
> Also, does anyone watch/check these? If so perhaps they can weigh in on what to call this and if a renaming would be overly disruptive.

oh yes, good catch @braingram, we should also change the `name:` field in the workflow to something more descriptive (for now `extra tests` would be consistent, I guess). That would mean the old `test on schedule` and `test with development versions` sub-tabs would still exist, but there would be a new `extra tests` sub-tab with the actual tests, so there wouldn't be any duplicate names. Let's talk about it at standup

File: .github/workflows/ci_cron.yml (hunk 6 above)
Comment (zacharyburnett) [reply .github/workflows/ci_cron.yml:70]: Didn't we want to make that another PR? Or maybe I'm thinking about something else we were talking about. I can make that change here, sure

**Comment by zacharyburnett** [comment]
> I'm also seeing "Cannot update this protected ref" on this and other PRs. Might that also be branch protection related?

That should be resolved as well now, try again? Sorry for the hassle, hopefully when this effort is done we can have a single branch protection that we never need to change

File: .github/workflows/ci_cron.yml (hunk 6 above)
Comment (braingram) [reply .github/workflows/ci_cron.yml:70]: I vote for as part of this PR.

**Comment by melanieclarke** [comment]
Fixed, thank you!  Merging is green again for other PRs.

File: .github/workflows/ci_cron.yml (hunk 6 above)
Comment (braingram) [reply .github/workflows/ci_cron.yml:70]: From offline discussion with @zacharyburnett he convinced me this is a large enough change to be put in a follow-up PR. @pllim does that work for you?

Comment (pllim) [reply .github/workflows/ci_cron.yml:70]: Sure. Just want to make sure we not overlooking it. Thanks!

File: .github/workflows/tests_extra.yml [hunk 7]
```diff
@@ -32,10 +32,10 @@ concurrency:
 
 jobs:
   latest_crds_contexts:
-    if: (github.repository == 'spacetelescope/jwst' && (github.event_name == 'schedule' || github.event_name == 'workflow_dispatch' || contains(github.event.pull_request.labels.*.name, 'run scheduled tests')))
+    if: (github.repository == 'spacetelescope/jwst' && (github.event_name == 'schedule' || github.event_name == 'workflow_dispatch' || contains(github.event.pull_request.labels.*.name, 'run extra tests')))
```
Comment (pllim) [inline .github/workflows/tests_extra.yml:35]: Circling back to label name. Since it is already renamed, no reason to keep changing it back and forth. Human readable name is fine. ID would be too opaque even if unique.

File: .github/workflows/tests_extra.yml [hunk 8]
```diff
@@ -59,10 +59,16 @@ jobs:
       artifact-path: |
         results.xml
       envs: |
+        - linux: check-docs-links
+        - macos: py311-xdist
+          pytest-results-summary: true
         - macos: py312-xdist
           pytest-results-summary: true
-        - linux: py311-xdist
+        - linux: py3-stdevdeps-xdist
```
Comment (pllim) [inline .github/workflows/tests_extra.yml:68]: Note for the next PR?
```suggestion
        # TODO: Merge stdevdeps into devdeps
        - linux: py3-stdevdeps-xdist
```

**Review by pllim — COMMENTED** [review]
Overall it is fine. I will let other maintainers look at the updated workflows as I don't have strong opinions what being tested there.

I would advise admin use admin override to merge *this* PR first before making changes to `main` branch protection rules.

**Review by braingram — APPROVED** [review]
LGTM once the tests finish.

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -1,36 +1,76 @@
...
-    rev: v1.17.1  # Use the latest stable version
+      - id: codespell
+        args:
+          - --write-changes
+          - --summary
+        additional_dependencies:
+          - tomli
+  # validate change log and fragments in `changes/`
+  - repo: https://github.com/twisted/towncrier
+    rev: 24.8.0
```
Comment (braingram) [inline .pre-commit-config.yaml]:
```suggestion
```
Let's not add this as part of this PR.

**Review by braingram — APPROVED** [review]
LGTM.

check-docs-links failure can't be reproduced locally (I can visit the failing url):
https://github.com/spacetelescope/jwst/actions/runs/17247088418/job/48939488924?pr=9770
oldestdeps failure is due to vo server timeout:
https://github.com/spacetelescope/jwst/actions/runs/17247088468/job/48939491839?pr=9770
The oldestdeps will need to be either rerun or this PR force merged as it's marked as required.
